File: algorithms/quicksort/quicksort.go
package main

import "fmt"

func Partition(arr []int, leftPointer int, rightPointer int) ([]int, int) {
	pivotIndex := rightPointer
	pivot := arr[pivotIndex]
	rightPointer--

	for {
		for arr[leftPointer] < pivot && leftPointer < len(arr) {
			leftPointer++
		}
		for arr[rightPointer] > pivot && rightPointer > 0 {
			rightPointer--
		}
		if leftPointer >= rightPointer {
			break
		} else {
			arr[leftPointer], arr[rightPointer] = arr[rightPointer], arr[leftPointer]
			leftPointer++
		}
	}
	arr[leftPointer], arr[pivotIndex] = arr[pivotIndex], arr[leftPointer]

	return arr, leftPointer
}

func Quicksort(arr []int, leftIndex int, rightIndex int) []int {
	if rightIndex > leftIndex {
		arr, pivotIndex := Partition(arr, leftIndex, rightIndex)
		arr = Quicksort(arr, leftIndex, pivotIndex-1)
		arr = Quicksort(arr, pivotIndex+1, rightIndex)
	}
	return arr
}

func main() {
	fmt.Println(Quicksort([]int{0, 1, 2, 3, 6, 5}, 0, 5))
	testArr := []int{55, 90, 74, 20, 16, 46, 43, 59, 2, 99, 79, 10, 73, 1, 68, 56, 3, 87, 40, 78, 14, 18, 51, 24, 57, 89, 4, 62, 53, 23, 93, 41, 95, 84, 88}
	fmt.Println(Quicksort(testArr, 0, len(testArr)-1))
}

File: algorithms/quicksort/quicksort.rb
class SortableArray
  attr_reader :array
  
  def initialize(array)
    @array = array
  end

  def partition!(left_pointer, right_pointer)
    # Right-most element as the pivot.
    pivot_index = right_pointer
  
    pivot = @array[pivot_index]
    right_pointer -= 1
    while true
      while @array[left_pointer] < pivot do
        left_pointer += 1
      end
      while @array[right_pointer] > pivot do
        right_pointer -= 1
      end
      if left_pointer >= right_pointer
        break
      else
        @array[left_pointer], @array[right_pointer] =
        @array[right_pointer], @array[left_pointer]
        left_pointer += 1
      end
    end
    @array[left_pointer], @array[pivot_index] =
    @array[pivot_index], @array[left_pointer]
    return left_pointer
  end

  def quicksort!(left_index, right_index)
    if right_index - left_index <= 0
      return
    end
    pivot_index = partition!(left_index, right_index)
    quicksort!(left_index, pivot_index - 1)
    quicksort!(pivot_index + 1, right_index)
  end
end

# array = [0, 5, 2, 1, 6, 3]

# sortable_array = SortableArray.new(array)
# sortable_array.quicksort!(0, array.length - 1)
# p sortable_array.array
